pilot/pkg/networking/core/v1alpha3/listener_builder.go

```diff
@@ -278,38 +259,84 @@ func (builder *ListenerBuilder) buildVirtualOutboundListener(
+
 	actualWildcard, _ := getActualWildcardAndLocalHost(node)
-	var listFilter []*listener.ListenerFilter
-	if util.IsAllowAnyOutbound(node) && node.SidecarScope.OutboundTrafficPolicy.EgressProxy != nil {
-		listFilter = append(listFilter, &listener.ListenerFilter{
-			Name: xdsutil.TlsInspector,
-		})
+
+	// Lets build the new listener. buildListenerOpts will allocate the filter chain slots
+	// with same length as the filterChainOptsArray. buildCompleteFilterChain will then populate these slots using
+	// the filter chains from the opts, as well as the ones added by the plugins.
+	listenerOpts := buildListenerOpts{
+		push:            push,
+		proxy:           node,
+		bind:            actualWildcard,
+		port:            int(push.Mesh.ProxyListenPort),
+		filterChainOpts: filterChainOptsArray,
+		bindToPort:      true,
+		skipUserFilters: true,
 	}
 
-	// add an extra listener that binds to the port that is the recipient of the iptables redirect
-	ipTablesListener := &xdsapi.Listener{
-		Name:             VirtualOutboundListenerName,
-		Address:          util.BuildAddress(actualWildcard, uint32(push.Mesh.ProxyListenPort)),
-		Transparent:      isTransparentProxy,
-		UseOriginalDst:   proto.BoolTrue,
-		ListenerFilters:  listFilter,
-		FilterChains:     filterChains,
-		TrafficDirection: core.TrafficDirection_OUTBOUND,
+	mutable := &plugin.MutableObjects{
+		Listener:     buildListener(listenerOpts),
+		FilterChains: getPluginFilterChain(listenerOpts),
+	}
+
+	// These are specific to the virtual outbound
+	mutable.Listener.TrafficDirection = core.TrafficDirection_OUTBOUND
+	mutable.Listener.Name = VirtualOutboundListenerName
+	mutable.Listener.Transparent = isTransparentProxy
+	mutable.Listener.UseOriginalDst = proto.BoolTrue
+
+	// Prepping to call the plugins
+	svc := util.FallThroughFilterChainBlackHoleService
+	if util.IsAllowAnyOutbound(node) {
+		svc = util.FallThroughFilterChainPassthroughService
 	}
-	ipTablesListener.ListenerFiltersTimeout = gogo.DurationToProtoDuration(push.Mesh.ProtocolDetectionTimeout)
-	if ipTablesListener.ListenerFiltersTimeout != nil {
-		ipTablesListener.ContinueOnListenerFiltersTimeout = true
+
+	redirectPort := &model.Port{
+		Port:     int(push.Mesh.ProxyListenPort),
+		Protocol: protocol.TCP,
+	}
+
+	listenerProtocol := plugin.ListenerProtocolTCP
+	if util.IsProtocolSniffingEnabledForOutbound(node) {
+		listenerProtocol = plugin.ListenerProtocolAuto
+	}
+	pluginParams := &plugin.InputParams{
+		ListenerProtocol:           plugin.ListenerProtocol(listenerProtocol),
+		DeprecatedListenerCategory: networking.EnvoyFilter_DeprecatedListenerMatch_SIDECAR_OUTBOUND,
+		Node:                       node,
+		Push:                       push,
+		Bind:                       actualWildcard,
+		Port:                       redirectPort,
+		Service:                    svc,
+	}
+	for _, p := range configgen.Plugins {
+		if err := p.OnOutboundListener(pluginParams, mutable); err != nil {
```

Comment:
Do we need to apply authn/authz plugin to virtual outbound listener?

Reply:
Nope. This is for the mixer telemetry filter

Reply:
But auth does do some mutation too

Reply:
```
// OnOutboundListener is called whenever a new outbound listener is added to the LDS output for a given service
// Can be used to add additional filters on the outbound path
func (Plugin) OnOutboundListener(in *plugin.InputParams, mutable *plugin.MutableObjects) error {
	if in.ServiceInstance == nil || in.Node.Type != model.Router {
		// Only care about router.
		return nil
	}

	return buildFilter(in, mutable)
}
```

```
// OnOutboundListener is called whenever a new outbound listener is added to the LDS output for a given service
// Can be used to add additional filters on the outbound path
func (Plugin) OnOutboundListener(in *plugin.InputParams, mutable *plugin.MutableObjects) error {
	if in.Node.Type != model.Router {
		// Only care about router.
		return nil
	}

	buildFilter(in, mutable)
	return nil
}
```

Reply:
It only cares about Router mode. The virtualOutbound is for sidecars only.

Reply:
My mistake